extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriLightHandler.java

```diff
@@ -0,0 +1,322 @@
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link TradfriLightHandler} is responsible for handling commands for
+ * individual lights.
+ *
+ * @author Kai Kreuzer - Initial contribution
+ */
+public class TradfriLightHandler extends BaseThingHandler implements CoapCallback {
+
+    private final Logger logger = LoggerFactory.getLogger(TradfriLightHandler.class);
+
+    // step size for increase/decrease commands
+    private static final int STEP = 10;
+
+    // keeps track of the current state for handling of increase/decrease
+    private LightData state = null;
+
+    // the unique instance id of the light
+    private Integer id;
+
+    // used to check whether we have already been disposed when receiving data asynchronously
+    private volatile boolean active = false;
+
+    private TradfriCoapClient coapClient;
+
+    public TradfriLightHandler(Thing thing) {
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        this.id = getConfigAs(DeviceConfig.class).id;
+        TradfriGatewayHandler handler = (TradfriGatewayHandler) getBridge().getHandler();
+        String uriString = handler.getGatewayURI() + "/" + id;
+        try {
+            URI uri = new URI(uriString);
+            coapClient = new TradfriCoapClient(uri);
+            coapClient.setEndpoint(handler.getEndpoint());
+        } catch (URISyntaxException e) {
+            logger.debug("Illegal device URI `{}`: {}", uriString, e.getMessage());
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
+            return;
+        }
+        updateStatus(ThingStatus.UNKNOWN);
+        active = true;
+        coapClient.startObserve(this);
+    }
+
+    @Override
+    public void dispose() {
+        active = false;
+        if (coapClient != null) {
+            coapClient.shutdown();
+        }
+        super.dispose();
+    }
+
+    @Override
+    public void setStatus(ThingStatus status, ThingStatusDetail statusDetail) {
+        if (active && getBridge().getStatus() != ThingStatus.OFFLINE && status != ThingStatus.ONLINE) {
+            updateStatus(status, statusDetail);
+        }
+    }
+
+    @Override
+    public void onUpdate(JsonElement data) {
+        if (active && !(data.isJsonNull())) {
+            state = new LightData(data);
+            updateStatus(state.getReachabilityStatus() ? ThingStatus.ONLINE : ThingStatus.OFFLINE);
```

Comment:
No, this is really just a flag - the rest of the functionality is kept in place: We can send commands and we receive updates - and if the gateway says that the light is reachable again, we simply set it to ONLINE again.